crates/core/tedge_mapper/src/sm_c8y_mapper/http_proxy.rs

```diff
@@ -0,0 +1,350 @@
+        -> Result<String, SMCumulocityMapperError>;
+}
+
+/// Define a C8y endpoint
+pub struct C8yEndPoint {
+    c8y_host: String,
+    device_id: String,
+    c8y_internal_id: String,
+}
+
+impl C8yEndPoint {
+    fn new(c8y_host: &str, device_id: &str, c8y_internal_id: &str) -> C8yEndPoint {
+        C8yEndPoint {
+            c8y_host: c8y_host.into(),
+            device_id: device_id.into(),
+            c8y_internal_id: c8y_internal_id.into(),
+        }
+    }
+
+    fn get_url_for_sw_list(&self) -> String {
+        let mut url_update_swlist = String::new();
+        url_update_swlist.push_str("https://");
+        url_update_swlist.push_str(&self.c8y_host);
+        url_update_swlist.push_str("/inventory/managedObjects/");
+        url_update_swlist.push_str(&self.c8y_internal_id);
+
+        url_update_swlist
+    }
+
+    fn get_url_for_get_id(&self) -> String {
+        let mut url_get_id = String::new();
+        url_get_id.push_str("https://");
+        url_get_id.push_str(&self.c8y_host);
+        url_get_id.push_str("/identity/externalIds/c8y_Serial/");
+        url_get_id.push_str(&self.device_id);
+
+        url_get_id
+    }
+
+    fn get_url_for_create_event(&self) -> String {
+        let mut url_create_event = String::new();
+        url_create_event.push_str("https://");
+        url_create_event.push_str(&self.c8y_host);
+        url_create_event.push_str("/event/events/");
+
+        url_create_event
+    }
+
+    fn get_url_for_event_binary_upload(&self, event_id: &str) -> String {
+        let mut url_event_binary = self.get_url_for_create_event();
+        url_event_binary.push_str(event_id);
+        url_event_binary.push_str("/binaries");
+
+        url_event_binary
+    }
+
+    fn url_is_in_my_tenant_domain(&self, url: &str) -> bool {
+        // c8y URL may contain either `Tenant Name` or Tenant Id` so they can be one of following options:
+        // * <tenant_name>.<domain> eg: sample.c8y.io
+        // * <tenant_id>.<domain> eg: t12345.c8y.io
+        // These URLs may be both equivalent and point to the same tenant.
+        // We are going to remove that and only check if the domain is the same.
+        let tenant_uri = &self.c8y_host;
+        let url_host = match Url::parse(url) {
+            Ok(url) => match url.host() {
+                Some(host) => host.to_string(),
+                None => return false,
+            },
+            Err(_err) => {
+                return false;
+            }
+        };
+
+        let url_domain = url_host.splitn(2, '.').collect::<Vec<&str>>();
+        let tenant_domain = tenant_uri.splitn(2, '.').collect::<Vec<&str>>();
+        if url_domain.get(1) == tenant_domain.get(1) {
+            return true;
+        }
+        false
+    }
+}
+
+/// An HttpProxy that uses MQTT to retrieve JWT tokens and authenticate the device
+///
+/// - Keep the connection info to c8y and the internal Id of the device
+/// - Handle JWT requests
+pub struct MqttAuthHttpProxy {
```

Comment:
Is the name actually right? Shouldn't that be something along the lines of `BearerAuthHttpProxy`?

Reply:
I'm okay with both name. One stressing the fact we use tokens, the other the way we retrieve the tokens.

Reply:
Changed to `JwtAuthHttpProxy`.